renku_notebooks/api/auth.py

```diff
@@ -0,0 +1,107 @@
+import json
+import logging
+import os
+import warnings
+from functools import wraps
+
+import requests
+from flask import Blueprint, current_app, jsonify, request, make_response, redirect, abort
+
+from jupyterhub.services.auth import HubOAuth
+
+from .. import config
+from ..util._kubernetes import annotate_servers
+
+try:
+    auth = HubOAuth(
+        api_token=os.environ.get('JUPYTERHUB_API_TOKEN'),
+        cache_max_age=60,
+    )
+except ValueError as e:
+    logging.warning(e)
+"""Wrap JupyterHub authentication service API."""
+
+bp = Blueprint('auth_bp', __name__, url_prefix=config.SERVICE_PREFIX)
+
+def authenticated(f):
+    """Decorator for authenticating with the Hub"""
+
+    @wraps(f)
+    def decorated(*args, **kwargs):
+        token = request.cookies.get(
+            auth.cookie_name
+        ) or request.headers.get('Authorization', '')[len('token'):].strip()
+        if token:
+            user = auth.user_for_token(token)
+        else:
+            user = None
+        if user:
+            return f(user, *args, **kwargs)
+        else:
+            # if the request is not coming from a browser, return 401
+            if request.is_xhr or request.environ['HTTP_ACCEPT'
+                                                 ] == 'application/json':
+                current_app.logger.info(
+                    'Unauthorized non-browser request - returning 401.'
+                )
+                response = jsonify(error='An authorization token is required.')
+                response.status_code = 401
+                return response
+
+            # redirect to login url on failed auth
+            state = auth.generate_state(next_url=request.path)
+            current_app.logger.debug(
+                'Auth flow, redirecting to: {}'.format(auth.login_url)
+            )
+            response = make_response(
+                redirect(auth.login_url + '&state=%s' % state)
+            )
+            response.set_cookie(auth.state_cookie_name, state)
+            return response
+
+    return decorated
+
+
+@bp.route('oauth_callback')
+def oauth_callback():
+    """Set a token in the cookie."""
+    code = request.args.get('code', None)
+    if code is None:
+        abort(403)
+
+    # validate state field
+    arg_state = request.args.get('state', None)
+    cookie_state = request.cookies.get(auth.state_cookie_name)
+    if arg_state is None or arg_state != cookie_state:
+        # state doesn't match
+        abort(403)
+
+    token = auth.token_for_code(code)
+    next_url = auth.get_next_url(cookie_state) or app.config.SERVICE_PREFIX
```

Comment:
Shouldn't this be `current_app.config.SER...`?

Reply:
yep! done